### ISSUE TITLE
TypeScript object is possibly undefined false positive

### ISSUE THREAD

**Issue by catalin-enache** [microsoft/TypeScript#41319]
**TypeScript Version:**  v4.0.3 or v4.1.0.dev.20201028 (Nightly)

**Search Terms:**
typescript object is possibly undefined false positive

**Code**

```ts
const myObject = { m1: "value", m2: 2 };
const entries: [string, string|number][] = Object.entries(myObject);
// const entries: [string, string|number][] = [['m1', 'value'], ['m2', 2]]; // same behavior
while (entries.length > 0) {
    const val = entries.shift();
    console.log(`The property name ${val[0]} has the value ${val[1]}`);
}
```

**Expected behavior:**
Transpile without errors.

**Actual behavior:**
Complains about 'val': Object is possibly 'undefined'.
A guard like: **val &&** console.log(`The property name ${val[0]} has the value ${val[1]}`);
is pointless/redundant since we're iterating while (entries.length > 0)

**Playground Link:** 

https://www.typescriptlang.org/play?target=99&ts=4.0.3#code/MYewdgzgLgBAtgTwPICMBWBTYsC8MDe8AjAFwwBEAbgIYA2ArhuQDTwBMZbMAvgNwCwAKFCRYGMFABOASwwQyAbWgywAc1bLpagD5h6cFBkkBdBcZh5UmbADpxU2RAAUiK1igBKAYID0PmCLQMPYycoqaahoOOnoGRqbmeAoKAORwRCmsKTQMGCnGrKlwbJkwbMbGvDB+MBDUcBgwhgAW1JTSIJJCAO7N0rSNTiGONgNqUM0wAHwwAAweBEIwywHgQTkWwRKhEDYQfQBmUE5eSyuBIAOjIKpOAAYAKs2NAA6SIC9GUAgwYPWNABJ8DkFLNjNwYK0IDAJo0cowYECQURwXdToJuEIgA

**Related Issues:** 

None so far.

**Comment by catalin-enache** [comment]
It seems like using "!" (val![0]/val![1]) is good enough solution which does not require an useless guard like **val &&**
